WordPress/WordPressTest/PagesListTests.swift

```diff
@@ -0,0 +1,502 @@
+    func testPerformance() throws {
+        // Make sure the total number of pages is about the same as the one in `testWorstPerformance`.
+        var pages = [Page]()
+        // Add pages whose parents are in the list.
+        for _ in 1...100 {
+            pages.append(contentsOf: parentPage(childrenCount: 5, additionalLevels: 4))
+        }
+        // Add pages whose parents are *not* in the list.
+        for _ in 1...80 {
+            let newPages = parentPage(childrenCount: 5, additionalLevels: 4)
+            newPages[0].parentID = NSNumber(value: randomID.next())
+            pages.append(contentsOf: newPages)
+        }
+        // Use a shuffled list to test performance, which in theory means more iterations in trying to find a page's parent page.
+        pages = pages.shuffled()
+        NSLog("\(pages.count) pages used in \(#function)")
+
+        measure {
+            let pageTree = PageTree()
+            pageTree.add(pages)
+            let list = pageTree.hierarchyList()
+            XCTAssertEqual(list.count, pages.count)
+        }
+    }
+
+    // Measure performance using a page list where contains non-top-level pages and none of their parent pages are in the list.
+    func testWorstPerformance() throws {
+        var pages = [Page]()
+        for id in 1...5000 {
+            let page = PageBuilder(mainContext).build()
+            page.postID = NSNumber(value: id)
+            page.parentID = NSNumber(value: randomID.next())
+            pages.append(page)
+        }
+        NSLog("\(pages.count) pages used in \(#function)")
+
+        measure {
+            let pageTree = PageTree()
+            pageTree.add(pages)
+            let list = pageTree.hierarchyList()
+            XCTAssertEqual(list.count, pages.count)
+        }
+    }
+
+    func testOrphanPagesNestedLists() throws {
+        var pages = [Page]()
+        let orphan1 = parentPage(childrenCount: 0)
+        pages.append(contentsOf: orphan1)
+        pages.append(contentsOf: parentPage(childrenCount: 2))
+        let orphan2 = parentPage(childrenCount: 2)
+        pages.append(contentsOf: orphan2)
+        pages.append(contentsOf: parentPage(childrenCount: 2))
+        orphan1[0].parentID = 100000
+        orphan2[0].parentID = 200000
+
+        makeAssertions(pages: pages)
+    }
+
+    func testHierachyListRepresentationRoundtrip() throws {
+        let roundtrip: (String) throws -> Void = { string in
+            let pages = try Array<Page>(hierarchyListRepresentation: string, context: self.mainContext)
+            XCTAssertEqual(PageTree.hierarchyList(of: pages).hierarchyListRepresentation(), string)
+        }
+
+        try roundtrip("""
+            1
+              2
+            3
+              4
+                5
+              6
+            7
+            """)
+
+        try roundtrip("""
+            1
+              2
+                3
+              4
+                5
+              6
+            7
+            """)
```

Comment:
This visual language to build the hierarchy is neat and useful.

Nitpick: Using two spaces indentation when the default for the Swift code set in the project is four makes it a bit odd. Have you considered using the same number of spaces?

Reply:
I feel like using 4 spaces is not as clear, in terms of seeing the hierachy... Here is an example
```
1
    2
3
    4
        5
    6
7
```